server/chat-plugins/net-filters.ts

```diff
@@ -192,35 +221,46 @@ export const PM = new QueryProcessManager<NetQuery, any>(module, async query =>
 		}
 		return 'success';
 	case 'trainfrom':
-		const {path: targetPath, result} = data as AnyObject;
-		const content = FS(targetPath).readSync();
+		const {path, result} = data as {path: string, result: string};
+		const content = FS(path).readSync();
 		const lines = NeuralNetChecker.sanitizeChatLines(content, result);
+		const length = lines.length;
 		const time = await net.train(lines);
-		return [time, lines.length];
+		return [time, length];
 	}
-}, PM_TIMEOUT);
+}
 
-if (!PM.isParentProcess) {
-	global.Config = Config;
+export const processes = {
+	// this one runs longer because training is SLOW
+	training: new ProcessManager.QueryProcessManager(module, handleQuery, PM_TIMEOUT),
+	// this one is fine
+	main: new ProcessManager.QueryProcessManager(module, handleQuery),
+}
 
-	global.Monitor = {
-		crashlog(error: Error, source = 'A netfilter process', details: AnyObject | null = null) {
-			const repr = JSON.stringify([error.name, error.message, source, details]);
-			process.send!(`THROW\n@!!@${repr}\n${error.stack}`);
-		},
-	};
-	process.on('uncaughtException', err => {
-		if (Config.crashguard) {
-			Monitor.crashlog(err, 'A net filter child process');
-		}
-	});
-	// we only want to spawn one network, when it's the subprocess
-	// otherwise, we use the PM for interfacing with the network
-	net = new NeuralNetChecker(PATH);
-	// eslint-disable-next-line no-eval
-	Repl.start('netfilters', cmd => eval(cmd));
-} else {
-	PM.spawn(NUM_PROCESSES);
+for (const k in processes) {
+	const PM = processes[k as keyof typeof processes];
+	if (!PM.isParentProcess) {
+		global.Config = Config;
+
+		global.Monitor = {
+			crashlog(error: Error, source = `A netfilter ${k} process`, details: AnyObject | null = null) {
+				const repr = JSON.stringify([error.name, error.message, source, details]);
+				process.send!(`THROW\n@!!@${repr}\n${error.stack}`);
+			},
+		};
+		process.on('uncaughtException', err => {
+			if (Config.crashguard) {
+				Monitor.crashlog(err, 'A net filter child process');
+			}
+		});
+		// we only want to spawn one network, when it's the subprocess
+		// otherwise, we use the PM for interfacing with the network
+		net = new NeuralNetChecker(PATH);
+		// eslint-disable-next-line no-eval
+		Repl.start(`netfilters-${k}-${process.pid}`, cmd => eval(cmd));
+	} else {
+		PM.spawn(NUM_PROCESSES[k as keyof typeof processes]);
+	}
 }
```

Comment:
```suggestion
const PM = processes.main;
const PMTraining = process.training;
if (!PM.isParentProcess) {
	global.Config = Config;

	global.Monitor = {
		crashlog(error: Error, source = `A netfilter ${k} process`, details: AnyObject | null = null) {
			const repr = JSON.stringify([error.name, error.message, source, details]);
			process.send!(`THROW\n@!!@${repr}\n${error.stack}`);
		},
	};
	process.on('uncaughtException', err => {
		if (Config.crashguard) {
			Monitor.crashlog(err, 'A net filter child process');
		}
	});
	// we only want to spawn one network, when it's the subprocess
	// otherwise, we use the PM for interfacing with the network
	net = new NeuralNetChecker(PATH);
	// eslint-disable-next-line no-eval
	Repl.start(`netfilters-${k}-${process.pid}`, cmd => eval(cmd));
} else {
	PM.spawn(NUM_PROCESSES.main);
	PMTraining.spawn(NUM_PROCESSES.training);
}
```

This fixes your double-response problem.

Reply:
In hindsight, this doesn't fully fix it, but it's a start.